web/src/stream_popover.js

```diff
@@ -214,7 +214,29 @@ function build_stream_popover(opts) {
     });
 }
 
-export function build_move_topic_to_stream_popover(
+async function get_message_type(current_stream_id, topic_name, current_message_id) {
+    const lastMessageId = await new Promise((resolve) => {
+        message_edit.with_last_message_id(current_stream_id, topic_name, (message_id) => {
+            resolve(message_id);
+        });
+    });
+
+    const firstMessageId = await new Promise((resolve) => {
+        message_edit.with_first_message_id(current_stream_id, topic_name, (message_id) => {
+            resolve(message_id);
+        });
+    });
+
+    if (current_message_id === firstMessageId) {
+        return "first_message";
+    } else if (current_message_id === lastMessageId) {
+        return "last_message";
+    }
+
+    return "intermediate_message";
+}
```

Comment:
So what I don't love about this is that it requires asking the server (and thus adding latency) when that should usually not be necessary. But we'll need this code for fallbacks in case locally cached data can't answer the question.

But, for example, if `current_msg_list.data.filter.supports_collapsing_recipients` and:
* If `current_msg_list.data` contains the message as well as messages both before and after the current message in the same conversation, then `intermediate` is the answer.
* If `current_msg_list.data.filter.is_conversation_view` and `.fetch_status.has_found_newest()`, and `.data.last()` is the message in question, we can return `"last_message"` without querying the server.
* If `current_msg_list.data.filter.is_conversation_view` and `.fetch_status.has_found_oldest()`, and `.data.first()` is the message in question, we can return `"first_message"` without querying the server.

Reply:
> If current_msg_list.data contains the message as well as messages both before and after the current message in the same conversation, then intermediate is the answer.

Would it be sufficient in this case to check only the adjacent messages, or do we have to traverse the whole `message_list` in order to verify if any message in the past or future might be of the same topic? That is the major doubt I have regarding this implementation. I have added an additional comment below explaining why I'm a bit confused by this.

Reply:
If the message list we're looking at is something like a stream narrow, then we have to traverse the whole message list potentially, as the closest adjacent messages might be in other topics.